feature: Reverse display order for "Open Tabs Above" option

laktak opened the pull request:
This is a new take on the "Open Tabs Above" option. Instead of inserting tabs before the current tab, only the display order is changed. The effect is that tabs are
automatically inserted in the correct position by Firefox. Also Ctrl+Tab works the same way as without TabCenter (by going to the new tab).

Fixes: #709
BREAKING CHANGE: Ctrl+Tab now selects tabs in the correct order.

ericawright commented:
thanks for the PR @laktak, I'll test #716 for you

laktak commented:
@ericawright That's strange, it works on OS X with Open Tabs Above enabled:

Before:

<img width="280" alt="t1" src="https://cloud.githubusercontent.com/assets/959858/19784299/883e7234-9c95-11e6-9a6f-2447a4a13d4e.png">

After force quit:

<img width="828" alt="t2" src="https://cloud.githubusercontent.com/assets/959858/19784300/884a04fa-9c95-11e6-8a1c-5003100596e3.png">

After restore:

<img width="273" alt="t3" src="https://cloud.githubusercontent.com/assets/959858/19784301/884f97e4-9c95-11e6-941e-c289eb46c019.png">

ericawright commented:
interesting... I tested by cmd-Q, rather than force quit. (in the bug, they both result in reversing order). I'll try force quit.

ericawright commented:
yup, you are correct. Force quit/crash will restore the tabs in the correct order. 
cmd-Q then `restore previous session` still restores in opposite order

laktak commented:
> cmd-Q then restore previous session still restores in opposite order

I don't get a "restore previous session" when I cmd-Q. Firefox starts up normally. Is that a setting somewhere? (I'm using the defaults in the developer edition)

ericawright commented:
yes, you click the History menu > Restore Previous Session

there is also a setting in `about:preferences` 
<img width="830" alt="screen shot 2016-10-27 at 5 03 56 pm" src="https://cloud.githubusercontent.com/assets/10803178/19785380/28d93a64-9c67-11e6-9753-551b4175e085.png">
If you have it set to `Show my windows and tabs from last time` then it's impossible to replicate. As they will restore on opening, and they always restore correctly.

laktak commented:
Got it, thanks :)

But even so, I can't repro this:

![oct-27-2016 23-12-17](https://cloud.githubusercontent.com/assets/959858/19785775/052ce244-9c9b-11e6-9afc-91e206304920.gif)

Firefox Developer 51.0a2 (2016-10-27) (64-bit)

ericawright commented:
hmm... I'm using nightly, I wonder if there's something else going on here.
![oct-27-2016 17-21-57](https://cloud.githubusercontent.com/assets/10803178/19785951/a296c9be-9c69-11e6-8fab-abf0e040ce6e.gif)

laktak commented:
I installed TabCenter 1.26 where I can repro the issue - not with my branch though.

ericawright commented:
@laktak, sorry about that - I was testing the right version from command line, and the wrong one with XPI 😳 

so, yes, your version does fix #716 which is really cool!
Some thoughts:
- Pinned tabs are now on the bottom. I'm not sure that's ideal.
- When changing the pref all the tabs move and reverse order - which could be positive or negative.
- When toggling top vs side tabs (the new buttons) if "open tabs above" is selected the order stays reversed - and new tabs etc. are opened to the left instead of right. (could be a feature? but I think we don't want to affect anything in top tabs mode)

laktak commented:
@ericawright no worries ;)

I fixed the pinning & top vs side issues. 

> When changing the pref all the tabs move and reverse order - which could be positive or negative, a little disorienting.

I think it has to work this way to avoid issues like [this](https://github.com/bwinton/TabCenter/issues/655#issuecomment-251502989).

laktak commented:
> When changing the pref all the tabs move and reverse order

Maybe we should change the name of the option though to "Newest Tabs on Top".

ericawright commented:
@phlsa can you test this XPI and let us know your thoughts? Specifically on the hotkey cmd-tab and cmd-shift-tab going in the opposite direction.
https://www.dropbox.com/s/ne5i2wu62a0j4hn/tabcentertest1%40mozilla.com-1.26.0.xpi?dl=0

phlsa commented:
Just discussed this with Erica -- let's do it!

ericawright commented:
@laktak sounds like we have approval. If you can rebase onto the latest master I'll merge it.
Also, if you could change the pref description that would be great. 
![screen shot 2016-11-01 at 2 18 18 pm](https://cloud.githubusercontent.com/assets/10803178/19901549/c7ec114c-a03d-11e6-888a-d00fc4ee60ef.png)
Adding a second line that says something like "Cmd-Tab and Ctrl-Tab will move through tabs in reverse direction"

bwinton commented:
Say, does this also change `Cmd-Shift-[`/`Cmd-Shift-]`?

ericawright commented:
@bwinton yes it does

laktak commented:
@ericawright great! I've rebased and changed the title & description.

The name of the option (`opentabstop`) is still the same. This might cause a surprise when the plugin updates (and all tabs are reordered) ... should we change this as well?

ericawright commented:
I believe we should keep it the same so it causes minimal disruption. As is, it should only affect users who have set the pref to true. 
I'll review and test once more soon.

ericawright commented on verticaltabs.js:
```diff
@@ -168,12 +144,27 @@ VerticalTabs.prototype = {
 
-    //reset _lastRelatedTab on changing preferences
+    let tabs = document.getElementById('tabbrowser-tabs');
+    let openTabsTop = !prefs.opentabstop;
```
this variable seems to never be actually used? Every time it is checked it seems `prefs.opentabstop` is also checked...

laktak commented:
@ericawright thanks for the feedback! I fixed the issues you mentioned.
